packages/navigator/src/browser/navigator-editor-sync.ts

```diff
@@ -0,0 +1,165 @@
+        }
+    }
+
+    private linkWithEditorPreferenceChanged(linkWithEditorNewValue: boolean | undefined) {
+        if (linkWithEditorNewValue) {
+            if (this.activeEditorChangedSubscription === undefined) {
+                this.activeEditorChangedSubscription = this.editorManager.onActiveEditorChanged(
+                    (editor: EditorWidget | undefined) => this.selectNodeByEditor(editor));
+            }
+        } else {
+            if (this.activeEditorChangedSubscription !== undefined) {
+                this.activeEditorChangedSubscription.dispose();
+                this.activeEditorChangedSubscription = undefined;
+            }
+        }
+    }
+
+    /**
+     * Reveals and selects corresponding to given editor node in the navigator.
+     * If editor is undefined nothing happens.
+     *
+     * @param editor editor widget object to reveal its node in the navigator
+     */
+    selectNodeByEditor(editor: EditorWidget | undefined) {
+        if (editor) {
+            this.selectNodeById(this.editorIdToNavigatorNodeId(editor.id));
+        }
+    }
+
+    /**
+     * Converts editor id to navigator node id.
+     * Example: 'code-editor-opener:file:///home/user/workspace/README.md' => 'file:///home/user/workspace/README.md'
+     *
+     * @param editorId id of editor tab
+     * @returns id of corresponding navigator node
+     */
+    private editorIdToNavigatorNodeId(editorId: string) {
+        return editorId.substring(editorId.indexOf(':') + 1);
+    }
+
+    /**
+     * Reveals and selects node in the navigator by node id.
+     * If node with given id doesn't exist, nothing happens.
+     * Node id example: 'file:///home/user/workspace/README.md'
+     *
+     * @param nodeId navigator tree node id
+     */
+    selectNodeById(nodeId: string): void {
+        let node = this.fileNavigatorModel.getNode(nodeId);
+        if (node) {
+            // node is mounted in the navigator tree
+            if (ISelectableTreeNode.is(node)) {
+                this.revealNode(node);
+                this.fileNavigatorModel.selectNode(node);
+            }
+        } else {
+            // node may exist, but hasn't mounted yet
+            this.tryRevealNode(nodeId).then((targetNode: ITreeNode | undefined) => {
+                if (targetNode && ISelectableTreeNode.is(targetNode)) {
+                    // node exists and revealed in the navigator tree now
+                    this.fileNavigatorModel.selectNode(targetNode);
+                }
+            });
+        }
+    }
+
+    /**
+     * Reveals given node in navigator.
+     * This method should be used when given node already exist (but might be hidden) in the navigator tree.
+     * Otherwise use this.tryRevealNode
+     */
+    private revealNode(node: ITreeNode | undefined): void {
+        if (node) {
+            // cannot use ITreeNode.isVisible(node) here because it says that node is visible when it actually isn't
```

Comment:
Why does it say so?

Reply:
I spotted it during debugging. Comment in the `ITreeNode` interface near `visible` field declares:
> /**
     * Test whether this node is visible.
     * If undefined then visible.
     */

But actually I spotted situation when it is undefined, but node isn't visible. So assumption `undefined === invisible` helped me.

Reply:
How can I reproduce it?

Reply:
Seems this is another bug, but this is my first PR here, so I didn't dare to change global model.

Reply:
Could you open an issue describing how to reproduce it?

Reply:
Ok, I'll try to reproduce it again